lib/model/security/AccountModel.class.php

```diff
@@ -86,47 +86,60 @@ public function maakAccount($uid) {
 	 * Verify SSHA hash.
 	 *
 	 * @param Account $account
-	 * @param string $pass_plain
+	 * @param string $passPlain
 	 * @return boolean
 	 */
-	public function controleerWachtwoord(Account $account, $pass_plain) {
-		$ohash = base64_decode(substr($account->pass_hash, 6));
+	public function controleerWachtwoord(Account $account, $passPlain) {
+		$valid = false;
+		// Controleer of het wachtwoord klopt
+		$hash = $account->pass_hash;
+		if (startsWith($hash, "{SSHA}")) {
+			$valid = $this->checkLegacyPasswordHash($passPlain, $hash);
+		} else {
+			$valid = password_verify($passPlain, $hash);
+		}
+
+		// Rehash wachtwoord als de hash niet aan de eisen voldoet
+		if ($valid && password_needs_rehash($hash, AccountModel::PASSWORD_HASH_ALGORITHM)) {
+			$this->wijzigWachtwoord($account, $passPlain, false);
+		}
+
+		return $valid === true;
+	}
+
+	private function checkLegacyPasswordHash($passPlain, $hash) {
+		$ohash = base64_decode(substr($hash, 6));
 		$osalt = substr($ohash, 20);
 		$ohash = substr($ohash, 0, 20);
-		$nhash = pack("H*", sha1($pass_plain . $osalt));
-		if ($ohash === $nhash) {
-			return true;
-		}
-		return false;
+		$nhash = pack("H*", sha1($passPlain . $osalt));
+		return $ohash == $nhash;
```

Comment:
Ik ben geen expert, maar toch even reactie op wat mij opviel.
Is een strict comparison niet meer robust? Dat was het trouwens eerst ook.
En allicht is het handig hier `hash_equals()` te gebruiken? 'Timing attack safe string comparison.'

Reply:
Ik zal hem inderdaad herschrijven naar hash_equals, om timing attacks te voorkomen.

Reply:
`hash_equals` is hier beter inderdaad, ook al is een timing attack niet feasible. Gezien het feit dat deze functie nog maar één keer per wachtwoord aangeroepen worden.